Dev/dronedome ahrs ekf

mdouglas90 opened the pull request:
This merges the mocap callback for the ahrs ekf. It does not change main so in order to use it, you must create LEQuad_dronedome not LEQuad and then call init (else the board will crash on startup). 

Word of caution: if the drone loses connection with the dronedome computers, it may fly into the ceiling or ground.

mdouglas90 commented:
Fixes #266

jlecoeur commented on sensing/ahrs_ekf.cpp:
```diff
@@ -394,7 +393,7 @@ Ahrs_ekf::Ahrs_ekf(const Imu& imu, ahrs_t& ahrs, const Ahrs_ekf::conf_t config):
 }
 
-bool Ahrs_ekf::update(void)
+bool Ahrs_ekf::predict_and_update(void)
```
What is the reason for this change ?
There should not be any name-clash issue with Kalman::update(z), if that is what you wanted to avoid.

jlecoeur replied:
The thing is we have update functions for all modules, so it would be a pity that this one is different.

jlecoeur commented on util/kalman.hpp:
```diff
@@ -296,7 +296,7 @@ class Kalman
 
     Mat<m,m> S_;    ///< Innovation covariance
-    Mat<n,n> I_;    ///< Identity matrix
+    const Mat<n,n> I_;    ///< Identity matrix
```
Why did you add `const` ?

jlecoeur replied:
Ok!

jlecoeur commented on sensing/ahrs_ekf_mocap.cpp:
```diff
@@ -0,0 +1,134 @@
+    callback.function       = (Mavlink_message_handler::msg_callback_func_t) &Ahrs_ekf_mocap::callback;
+    callback.module_struct  = (Mavlink_message_handler::handling_module_struct_t)        this;
+
+    message_handler.add_msg_callback(&callback);
```
I would not recommend this as this assumes that message_handler is fully initialised (ie. memory for callback list allocated) when this constructor is called. 
It is not the case if Ahrs_ekf is declared before mavlink_communication in the MAV class.

Although it would be nicer to do it like you did, we have to add an init function and call it after all the objects are created.

mdouglas90 commented:
I changed the name of the function to predict_and_update to reduce the ambiguity between `Kalman::update(...)` and `Ahrs_ekf::update()`, even though they won't clash. It's also a little misleading because the old update function was calling the predict function and then update function. We can change it back if you'd like.

I added const to I_ cause I had it as const in my Ahrs_ekf_mocap class and it's an identity matrix so it should never change.

jlecoeur commented on sensing/ahrs_ekf_mocap.hpp:
```diff
@@ -0,0 +1,124 @@
+    Mat<7,7>& P_;       ///< State covariance
+    const Mat<7,7> I_;  ///< Identity matrix
+    Mat<4,4> R_mocap_;  ///< The mocap measurement noise matrix
+
```
It seems a bit convoluted to take references to some matrices of ahrs_ekf, and to reallocate memory for an identity matrix that is already in the ahrs_ekf class.

I would have taken a reference to ahrs_ekf. Kept only R and H in this class. Then in the callback simply done:
`ahrs_ekf_.update(z, H, R);`

jlecoeur commented:
The thing is we have update functions for all modules, so it would be a pity that this one is different.

jlecoeur commented:
About your first message:
- why would the board crash if you do not initialize Ahrs_ekf_mocap?
- why would it got to the ceiling if it lost connexion?

jlecoeur commented on sensing/ahrs_ekf.cpp:
```diff
@@ -472,3 +471,13 @@ bool Ahrs_ekf::update(void)
+Mat<7,7>& Ahrs_ekf::P()
+{
+    return P_;
+}
```
This is only needed because you take non-const references to these matrices in ahrs_ekf_mocap. I would rather remove this as
- it breaks encapsulation
- the const version are already defined in the base class `const Mat<n,1>& Kalman::x() const` and `const Mat<n,n>& Kalman::P() const`

jlecoeur commented on sensing/ahrs_ekf_mocap.cpp:
```diff
@@ -0,0 +1,134 @@
+    z(2, 0) = packet.q[2];
+    z(3, 0) = packet.q[3];
+
+    Mat<4, 7> H = Mat<4, 7>(0.0f);
```
z and H could be members of ahrs_ekf_mocap (it avoids creating them 10 times per second)

mdouglas90 commented:
> why would the board crash if you do not initialize Ahrs_ekf_mocap?

I noticed a while ago (before I started working on the ahrs mocap code) that the autopilot crashed on startup if `LEQuad_dronedome::init()` was not called. This does not seem to be the case anymore. Although `LEQuad_dronedome::init()` must still be called for LEQuad_dronedome in order to receive the mocap messages

> why would it got to the ceiling if it lost connexion?

When the drone is not receiving messages from the optitrack system, it's position estimation starts to drift and can diverge quickly (or at least is has been doing on my drone). If in hold position, for example, the drone would try to compensate for its perceived change in position and move accordingly.
